ci_framework/plugins/action/ci_script.py

```diff
@@ -0,0 +1,150 @@
+        required: true
+        type: str
+    debug:
+        description: Enables bash tracing in the script
+        required: true
+        type: bool
+'''
+
+EXAMPLES = r'''
+- name: Run custom script
+  register: script_output
+  ci_script:
+    output_dir: "/home/zuul/ci-framework-data/artifacts"
+    cmd: |
+      mkdir /home/zuul/test-dir
+      cd /home/zuul/test-dir
+      git clone https://github.com/openstack-k8s-operators/ci-framework.git
+'''
+
+RETURN = r'''
+changed:
+   description: Always true.
+   returned: always
+   type: bool
+failed:
+   description: True if the execution failed.
+   returned: always
+   type: bool
+rc:
+   description: Script return code.
+   returned: always
+   type: int
+stderr:
+    description: STDERR output as string.
+    returned: always
+    type: str
+stderr_line:
+    description: STDERR output as lines.
+    returned: always
+    type: list[str]
+stdout:
+    description: STDOUT output as string.
+    returned: always
+    type: str
+stdout_line:
+    description: STDOUT output as lines.
+    returned: always
+    type: list[str]
+'''
+
+import glob
+import pathlib
+import re
+import uuid
+
+from ansible.plugins.action import ActionBase
+from ansible.errors import AnsibleActionFail
+
+
+TMPL_SCRIPT = '''#!/bin/bash
+%(opts)s
+exec > >(tee -ia %(logpath)s)
+%(content)s
+'''
+
+class ActionModule(ActionBase):
+
+    def __init__(self, **kwargs):
+        super(ActionModule, self).__init__(**kwargs)
+        self.__script_file_path = pathlib.Path('~/.ansible/tmp').joinpath(uuid.uuid4().hex).absolute()
+        self.__script_file_path.parent.mkdir(parents=True, exist_ok=True)
+
+    @staticmethod
+    def __build_options(task_args):
+        opts = 'set -euo pipefail'
+        if task_args.get('debug', True):
+            opts += f'\nset -x'
+        return opts
+
+    def run(self, tmp=None, task_vars=None):
+
+        super(ActionModule, self).run(tmp, task_vars)
+
+        if 'output_dir' not in self._task.args:
+            raise AnsibleActionFail('output_dir parameter is missing')
+
+        if 'script' not in self._task.args:
+            raise AnsibleActionFail('script parameter is missing')
+
+        output_dir = pathlib.Path(self._task.args['output_dir'])
+
+        fnum = len(glob.glob(f'{output_dir}/ci_script_*'))
+        t_name = re.sub(r'([^\x00-\x7F]|\s)+', '_', self._task._name).lower()
+        log_path = output_dir.parent.joinpath('logs', f'ci_script_{fnum:03}_{t_name}.log')
+        script_template_data = {
+           'content': self._task.args['script'],
+           'logpath': log_path.as_posix(),
+           'opts': self.__build_options(self._task.args)
+        }
+
+        script_path_str = self.__script_file_path.as_posix()
+        with open(self.__script_file_path, 'w') as tmp_script_file:
+            script_content = TMPL_SCRIPT % script_template_data
+            tmp_script_file.write(script_content)
+
+        file_task = self._task.copy()
```

Comment:
sooo here, you're basically passing the same parameters to the ansible.builtin.script module - and since you didn't remove the `output_dir` and `script` parameters, it will fail because of unknown parameters. Check what was done in the `ci_make` action plugin.

Reply:
Well, maybe this is part of Ansible not being consistent, but this works, there is no error regarding unknown parameters, seems that every action/module is designed differently. I can change it to be consistent anyway.